packages/snaps-controllers/src/websocket/WebSocketService.ts

```diff
@@ -0,0 +1,254 @@
+  HandleSnapRequest,
+  SnapInstalled,
+  SnapUninstalled,
+  SnapUpdated,
+} from '../snaps';
+import { METAMASK_ORIGIN } from '../snaps';
+
+const serviceName = 'WebSocketService';
+
+export type OpenWebSocket = {
+  type: `${typeof serviceName}:open`;
+  handler: WebSocketService['open'];
+};
+
+export type CloseWebSocket = {
+  type: `${typeof serviceName}:close`;
+  handler: WebSocketService['close'];
+};
+
+export type SendWebSocketMessage = {
+  type: `${typeof serviceName}:sendMessage`;
+  handler: WebSocketService['sendMessage'];
+};
+
+export type GetAll = {
+  type: `${typeof serviceName}:getAll`;
+  handler: WebSocketService['getAll'];
+};
+
+export type WebSocketServiceActions =
+  | OpenWebSocket
+  | CloseWebSocket
+  | SendWebSocketMessage
+  | GetAll;
+
+export type WebSocketServiceAllowedActions = HandleSnapRequest;
+
+export type WebSocketServiceEvents =
+  | SnapUninstalled
+  | SnapUpdated
+  | SnapInstalled;
+
+export type WebSocketServiceMessenger = RestrictedMessenger<
+  'WebSocketService',
+  WebSocketServiceActions | WebSocketServiceAllowedActions,
+  WebSocketServiceEvents,
+  WebSocketServiceAllowedActions['type'],
+  WebSocketServiceEvents['type']
+>;
+
+type WebSocketServiceArgs = {
+  messenger: WebSocketServiceMessenger;
+};
+
+type InternalSocket = {
+  id: string;
+  snapId: SnapId;
+  url: string;
+  openPromise: Promise<void>;
+  // eslint-disable-next-line no-restricted-globals
+  socket: WebSocket;
+};
+
+export class WebSocketService {
+  readonly #messenger: WebSocketServiceMessenger;
+
+  readonly #sockets: Map<string, InternalSocket>;
+
+  constructor({ messenger }: WebSocketServiceArgs) {
+    this.#messenger = messenger;
+    this.#sockets = new Map();
+
+    this.#messenger.registerActionHandler(
+      `${serviceName}:open`,
+      this.open.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${serviceName}:close`,
+      this.close.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${serviceName}:sendMessage`,
+      this.sendMessage.bind(this),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${serviceName}:getAll`,
+      this.getAll.bind(this),
+    );
+
+    this.#messenger.subscribe('SnapController:snapUpdated', (snap) => {
+      this.closeAll(snap.id);
+    });
+
+    this.#messenger.subscribe('SnapController:snapUninstalled', (snap) => {
+      this.closeAll(snap.id);
+    });
+
+    // Due to local Snaps not currently triggering uninstalled we also close connections for new Snaps.
+    this.#messenger.subscribe('SnapController:snapInstalled', (snap) => {
+      this.closeAll(snap.id);
+    });
+  }
+
+  #get(snapId: SnapId, id: string) {
+    const socket = this.#sockets.get(id);
+
+    assert(
+      socket && socket.snapId === snapId,
+      `Socket with ID "${id}" not found.`,
+    );
+
+    return socket;
+  }
+
+  #exists(snapId: SnapId, url: string) {
+    return this.getAll(snapId).some((socket) => socket.url === url);
```

Comment:
Wondering if this should be unique based on the combination of URL and protocols.